CONTRIBUTING.md

```diff
@@ -15,39 +15,173 @@ Otherwise, go read Docker's
+#### Fork the Swarm repo
+
+You'll want to create a personal fork of the Swarm repo.  This allows you to make changes in your own area and then later open pull requests to propose your work to the main Swarm project.
+
+Go to the [`docker/swarm` repo](https://github.com/docker/swarm) in a browser and click the "Fork" button in the top right corner.  You'll end up with your form of Swarm in `https://github.com/<username>/swarm`.
+
+(Throughout this document, `<username>` refers to your Github username.)
+
+#### Clone Swarm and set remotes
+
+Now clone the main Swarm repo:
+
+```sh
+mkdir -p $GOPATH/src/github.com/docker
+cd $GOPATH/src/github.com/docker
+git clone https://github.com/docker/swarm.git
+cd swarm
+```
+
+For the easiest contribution workflow, we will set up remotes as follows:
+  * `origin` is your personal fork
+  * `upstream` is `docker/swarm`
+
+Run these commands:
+
+```sh
+git remote remove origin
+git remote add origin https://github.com/<username>/swarm.git
+git remote add upstream https://github.com/docker/swarm.git
+git remote set-url --push upstream no-pushing
+```
+
+You can check your configuration like this:
 ```sh
-go get -d github.com/docker/swarm
+$ git remote -v
+origin     https://github.com/mgoelzer/swarm.git (fetch)
+origin     https://github.com/mgoelzer/swarm.git (push)
+upstream   https://github.com/docker/swarm.git (fetch)
+upstream   no-pushing (push)
 ```
 
-#### Start hacking
+#### Pull `docker/swarm` changes into your fork
+
+As the `docker/swarm` moves forward, it will accumulate commits that are not in your personal fork.  To keep your fork up to date, you need to periodically rebase:
 
-First, prepare dependencies for `swarm` and try to compile it.
 ```sh
-cd src/github.com/docker/swarm
-$GOBIN/godep restore
-go test -v -race ./...
-go install
+git fetch upstream master
+git rebase upstream/master
+git push -f origin master
 ```
 
-#### Adding New Dependencies
+Don't worry:  fetching the upstream and rebasing will not overwrite your local changes.
+
+### Build the Swarm binary
+
+Build the binary, installing it to `$GOPATH/bin/swarm`:
+
+```sh
+cd $GOPATH/src/github.com/docker/swarm
+godep go install .
+```
+ 
+Run the binary you just created:
+
+```sh
+$GOPATH/bin/swarm help
+$GOPATH/bin/swarm create
+$GOPATH/bin/swarm manage token://<cluster_id>
+$GOPATH/bin/swarm join --addr=<node_ip:2375> token://<cluster_id>
+```
+
+#### Create a Swarm image for deployment
+
+Swarm is distributed as a Docker image with a single `swarm` binary inside.  To create the image:
+
+```sh
+docker build -t my_swarm .
+```
+
+Now you can run the same commands as above using the container:
+
+```sh
+docker run my_swarm help
+docker run -d my_swarm join --addr=<node_ip:2375> token://<cluster_id>
+docker run -d -p <manager_port>:2375 my_swarm manage token://<cluster_id>
+```
+
+For complete documentation on how to use Swarm, refer to the Swarm section of [docs.docker.com](http://docs.docker.com/).
+
```

Comment:
Unit nit test.
```
godep go test -race ./...
```

Integration test

```
./test/integration/run.sh
```

Use this command to check if *.bats files are formatted correctly; otherwise it'll fail docker/dco-signed.
```
find test/ -type f \( -name "*.sh" -or -name "*.bash" -or -name "*.bats" \) -exec grep -Hn -e "^ " {} \;
```